src/llvm/analysis/ReachingDefinitions/ReachingDefinitions.h

```diff
@@ -3,141 +3,46 @@
-
-    struct Subgraph {
-        Subgraph(RDNode *r1, RDNode *r2)
-            : root(r1), ret(r2) {}
-        Subgraph(): root(nullptr), ret(nullptr) {}
-
-        RDNode *root;
-        RDNode *ret;
-    };
-
-    // points-to information
-    dg::LLVMPointerAnalysis *PTA;
-
-    // map of all nodes we created - use to look up operands
-    std::unordered_map<const llvm::Value *, RDNode *> nodes_map;
-
-    // mapping of llvm nodes to relevant reaching definitions nodes
-    // (this is a super-set of nodes_map)
-    // we could keep just one map of these two and don't duplicate
-    // the information, but this way it is more bug-proof
-    std::unordered_map<const llvm::Value *, RDNode *> mapping;
-
-    // map of all built subgraphs - the value type is a pair (root, return)
-    std::unordered_map<const llvm::Value *, Subgraph> subgraphs_map;
-    // list of dummy nodes (used just to keep the track of memory,
-    // so that we can delete it later)
-    std::vector<RDNode *> dummy_nodes;
-public:
-    LLVMRDBuilder(const llvm::Module *m,
-                  dg::LLVMPointerAnalysis *p,
-                  bool pure_funs = false)
-        : M(m), DL(new llvm::DataLayout(m)),
-          assume_pure_functions(pure_funs), PTA(p) {}
-    ~LLVMRDBuilder();
-
-    RDNode *build();
-
-    // let the user get the nodes map, so that we can
-    // map the points-to informatio back to LLVM nodes
-    const std::unordered_map<const llvm::Value *, RDNode *>&
-                                getNodesMap() const { return nodes_map; }
-    const std::unordered_map<const llvm::Value *, RDNode *>&
-                                getMapping() const { return mapping; }
-
-    RDNode *getMapping(const llvm::Value *val)
-    {
-        auto it = mapping.find(val);
-        if (it == mapping.end())
-            return nullptr;
-
-        return it->second;
-    }
-
-    RDNode *getNode(const llvm::Value *val)
-    {
-        auto it = nodes_map.find(val);
-        if (it == nodes_map.end())
-            return nullptr;
-
-        return it->second;
-    }
-
-    RDNode *getOperand(const llvm::Value *val);
-    RDNode *createNode(const llvm::Instruction& Inst);
-
-private:
-    void addNode(const llvm::Value *val, RDNode *node)
-    {
-        auto it = nodes_map.find(val);
-        assert(it == nodes_map.end() && "Adding a node that we already have");
-
-        nodes_map.emplace_hint(it, val, node);
-        node->setUserData(const_cast<llvm::Value *>(val));
-    }
-
-    ///
-    // Add a dummy node for which there's no real LLVM node
-    void addNode(RDNode *node)
-    {
-        dummy_nodes.push_back(node);
-    }
-
-    void addMapping(const llvm::Value *val, RDNode *node)
-    {
-        auto it = mapping.find(val);
-        assert(it == mapping.end() && "Adding mapping that we already have");
-
-        mapping.emplace_hint(it, val, node);
-    }
+namespace detail {
+    template <typename Rda>
+        struct BuilderSelector {
+            // TODO dense
```

Comment:
What does this TODO mean?

Reply:
I've used something as a placeholder on the line below it, so it was just a reminder to replace it by `LLVMRDBuilderDense`, which is already in place. I just forgot to remove it :)